### ISSUE TITLE
a few pending submissions make other completions hangs forever in kernel 5.4 lts

### ISSUE THREAD

**Issue by omegacoleman** [axboe/liburing#205]
I noticed that, when submitted a few reads that take long time to complete, those submissions may "hang" the other unrelated submissions from getting their CQEs -- the wait never returns unless those reads complete. This only happens in 5.4 and seem fixed in 5.8 & linux-next, but could still be a problem as 5.4 is picked as lts.
At first glance it seems like the CQ overflow bug, but I'm sure it is not because I've raised the length up to 1024. Plus the CQEs were not "dropped" -- they'll return alongside the pending ones if they ever complete.
I could repr with following code, it's a bit of c++ but should be easier to understand, the demonstration uses 2 evfd, but the problem exists with any type of fd actually.

```c++
#include <sys/uio.h>
#include <sys/eventfd.h>
#include <unistd.h>
#include <liburing.h>
#include <iostream>

void exit_perror(const char *s)
{
  perror(s);
  std::terminate();
}

void read_from_evfd(io_uring *ring, int fd)
{
  int64_t *buf = new int64_t;
  auto *iov = new iovec;
  iov->iov_base = static_cast<void *>(buf);
  iov->iov_len = sizeof(int64_t);
  auto *sqe = io_uring_get_sqe(ring);
  io_uring_prep_readv(sqe, fd, iov, 1, 0);
}

void sync_write_evfd(int fd)
{
  int64_t buf = 1;
  if (write(fd, &buf, sizeof(int64_t)) < 0)
    exit_perror("write");
}

int main(void)
{
  int evfd1 = eventfd(0, 0);
  int evfd2 = eventfd(0, 0);
  io_uring ring;
  io_uring_queue_init(1024, &ring, 0);
  for (;;)
  {
    read_from_evfd(&ring, evfd1);  // first submit lotta blocking workload on one fd
    read_from_evfd(&ring, evfd1);
    read_from_evfd(&ring, evfd1);
    read_from_evfd(&ring, evfd1);
    read_from_evfd(&ring, evfd1);
    read_from_evfd(&ring, evfd1);

    read_from_evfd(&ring, evfd2);  // then another workload that completes immediatly
    sync_write_evfd(evfd2);

    io_uring_submit_and_wait(&ring, 1);  // never returns with 5.4.x, returns on 5.8 & linux-next

    io_uring_cqe *cqe;
    if (io_uring_peek_cqe(&ring, &cqe) < 0)
      exit_perror("io_uring_peek_cqe");
    if (cqe->res < 0)
      exit_perror("cqe->res");
    io_uring_cqe_seen(&ring, cqe);

    std::cout << "#";
    std::cout.flush();
  }
  io_uring_queue_exit(&ring);
  return 0;
}
```

running this with kernel 5.8 & linux-next result in lots of "#"s -- as intended;
but with kernel 5.4.33 vanila and 5.4.55-hwe (shipped by ubuntu), the program stucks at io_uring_submit_and_wait forever

**Comment by YoSTEALTH** [comment]
@omegacoleman I think the solution is to get the linux distro you are using to backport latest io_uring to 5.4 or do this yourself, (so I have herd).

**Comment by omegacoleman** [comment]
> @omegacoleman I think the solution is to get the linux distro you are using to backport latest io_uring to 5.4 or do this yourself, (so I have heard).

i could use newer kernel for myself, but 5.4 is marked as lts

**Comment by isilence** [comment]
Not really. We just need to find out what causing this and backport some patches for the next stable release of 5.4+

**Comment by isilence** [comment]
I suspect that's fixed by introducing a new internal threadpool (io-wq).
IMHO, that's not a thing that can be backported (too large + fixes on top).
So, I'm afraid that's set in stone, if Jens don't have another opinion.

**Comment by omegacoleman** [comment]
> I suspect that's fixed by introducing a new internal threadpool (io-wq).
> IMHO, that's not a thing that can be backported (too large + fixes on top).
> So, I'm afraid that's set in stone, if Jens don't have another opinion.

so the kernel got a new threadpool now -- that's very interesting, i'm looking into the patches

however I think this could be a bug in the workqueue io-wq replaced -- this behaivor is even no way to workaround, because it's unstable

besides this could be a huge barrier for anyone (including me) trying to propagate io_uring -- because there would be lots of people using the lts kernel. imagine you distributed some program, and it got stuck on their machines in the middle of nowhere, and there's no way out but to upgrade their kernel

so pls have a look @axboe

**Comment by isilence** [comment]
I'll probe into it a bit more

**Comment by seemk** [comment]
I think I stumbled upon the same issue, the relevant code is here: https://github.com/seemk/urn/blob/master/uring/relay.cpp#L389-L404

If I initially submit a lot of reads there, the ring will hang and it was driving me up the wall :disappointed:

**Comment by xjtuwjp** [comment]
Any update, I can easily reproduce the problem with 5.4.95, I see there is no change since in latest 5.4.99

**Comment by isilence** [comment]
My bad, forgot to reply. It's too old, not sure stable even accepts patches there.
Especailly that it's not really a bug but a restirction of old kernel versions.

Looking at the code, I guess the reason is explicit restriction of # of workers.
```C
max_active = min(ctx->sq_entries - 1, 2 * num_online_cpus()));
ctx->sqo_wq[0] = alloc_workqueue("io_ring-wq", WQ_UNBOUND | WQ_FREEZABLE, max_active;

max_active = 2;
ctx->sqo_wq[1] = alloc_workqueue("io_ring-write-wq", WQ_UNBOUND | WQ_FREEZABLE, max_active);
```

```
``@max_active`` determines the maximum number of execution contexts
per CPU which can be assigned to the work items of a wq.  For example,
with ``@max_active`` of 16, at most 16 work items of the wq can be
executing at the same time per CPU.
```

**Comment by xjtuwjp** [comment]
Thanks @isilence  for the update.  it this problem only limited to eventfd or normal filefd?

**Comment by isilence** [comment]
> Thanks @isilence for the update. it this problem only limited to eventfd or normal filefd?

Not limited to evenfd, but about all blocking I/O.
E.g. direct block device read/write shouldn't be affected, but buffered I/O, sockets, pipes, etc. will be.

**Comment by xjtuwjp** [comment]
Thanks for the quick reply. @isilence , luckily our use case is only direct block device IO via QEMU.

**Comment by isilence** [comment]
TL;DR; the issue is for a very old kernel due to io_uring being not feature complete at that point.
Can't be fixed upstream, let's close it, maybe?

**Comment by beef9999** [comment]
@isilence Is it true that we should avoid using Kernel 5.4 to run io_uring?

Perhaps some documents should be made. Considering 5.4 is still the long-term and regular developers only have the idea that io_uring appear in kernel 5.1

**Comment by beef9999** [comment]
> @isilence @axboe Is it true that we should avoid using Kernel 5.4 to run io_uring?
> 
> Perhaps some documents should be made. Considering 5.4 is still the long-term. Regular developers only have the idea that io_uring appears in kernel 5, and for now 5.4 is the only 5.x LT version that failed to support io_uring...

I rerun some tests in 5.4, and found the IO would fail with errno 22(Invalid argument). Looks like someone has disabled it from kernel... That would be fine then...

**Comment by isilence** [comment]
> @isilence @axboe Is it true that we should avoid using Kernel 5.4 to run io_uring?
> 
> Perhaps some documents should be made. Considering 5.4 is still the long-term. Regular developers only have the idea that io_uring appears in kernel 5, and for now 5.4 is the only 5.x LT version that failed to support io_uring...

It's certainly recommended to take newer kernels, just because 5.4 was long ago and lacks lots of features io_uring-wise that came after.

**Comment by isilence** [comment]
> I rerun some tests in 5.4, and found the IO would immediately fail with errno 22(Invalid argument). Looks like someone has disabled it from kernel... That would be fine then...

Didn't get it, what is disabled? io_uring?
Is it only IO or  syscall__io_uring_setup (e.g. io_uring_queue_init()) fails as well?
